lib/internal/modules/esm/resolve.js

```diff
@@ -153,44 +175,15 @@ function fileExists(url) {
  * @returns {URL}
  */
 function legacyMainResolve(packageJSONUrl, packageConfig, base) {
-  let guess;
-  if (packageConfig.main !== undefined) {
-    // Note: fs check redundances will be handled by Descriptor cache here.
-    if (fileExists(guess = new URL(`./${packageConfig.main}`,
-                                   packageJSONUrl))) {
-      return guess;
-    } else if (fileExists(guess = new URL(`./${packageConfig.main}.js`,
-                                          packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}.json`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}.node`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}/index.js`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}/index.json`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}/index.node`,
-                                        packageJSONUrl)));
-    else guess = undefined;
-    if (guess) {
-      emitLegacyIndexDeprecation(guess, packageJSONUrl, base,
-                                 packageConfig.main);
-      return guess;
-    }
-    // Fallthrough.
-  }
-  if (fileExists(guess = new URL('./index.js', packageJSONUrl)));
-  // So fs.
-  else if (fileExists(guess = new URL('./index.json', packageJSONUrl)));
-  else if (fileExists(guess = new URL('./index.node', packageJSONUrl)));
-  else guess = undefined;
-  if (guess) {
-    emitLegacyIndexDeprecation(guess, packageJSONUrl, base, packageConfig.main);
-    return guess;
-  }
-  // Not found.
-  throw new ERR_MODULE_NOT_FOUND(
-    fileURLToPath(new URL('.', packageJSONUrl)), fileURLToPath(base));
+  const baseStringified = isURL(base) ? base.href : base;
```

Comment:
Isn't that more performant?
```suggestion
  const baseStringified = `${base}`;
```

Reply:
But `base` can be `undefined`, so it will be stringified as `'undefined'` and this will not lead to the same error as the previous version.

Reply:
I think this matches the previous version and is perhaps slightly faster?

```suggestion
  const baseStringified = base?.href;
```

Reply:
Accords the method signature, `base` can be a string too, the alternative could be `base?.href ?? base`, but I think checking using `isURL` will be faster and more reliable.

Reply:
Could we simplify the method signature so that `base` is always passed in as either a string or as a URL, rather than accepting either? I would think that eliminating the `isURL` call should make this slightly faster.

Reply:
`base` is used as a URL in many other places, converting it to a string will not be good for performance, so I will keep it as is.